gibbon-compiler/src/Gibbon/Passes/CalculateBounds.hs

```diff
@@ -0,0 +1,109 @@
+module Gibbon.Passes.CalculateBounds where
+
+import           Gibbon.Common
+import qualified Data.Map                      as M
+import           Gibbon.L2.Syntax
+import           Debug.Trace
+import           Data.Maybe
+
+
+
+calculateBounds :: Prog2 -> PassM Prog2
+calculateBounds Prog { ddefs, fundefs, mainExp } = do
+  let env2 = Env2 M.empty (initFunEnv fundefs)
+  fundefs' <- mapM (calculateBoundsFun ddefs env2 M.empty) fundefs
+  mainExp' <- case mainExp of
+    Nothing       -> return Nothing
+    Just (mn, ty) -> Just . (, ty) . fst <$> calculateBoundsExp2 ddefs env2 M.empty mn
+  return $ Prog ddefs fundefs' mainExp'
+
+
+calculateBoundsFun :: DDefs Ty2 -> Env2 Ty2 -> M.Map Var RegionSize -> FunDef2 -> PassM FunDef2
+calculateBoundsFun ddefs env2 szEnv f@FunDef { funBody } = do
+  funBody' <- fst <$> calculateBoundsExp2 ddefs env2 szEnv funBody
+  return $ f { funBody = funBody' }
+
+calculateBoundsExp2 ddefs env2 szEnv ex = do
+  res <- calculateBoundsExp ddefs env2 szEnv ex
+  traceShowM res
+  return res
+
+calculateBoundsExp :: DDefs Ty2 -> Env2 Ty2 -> M.Map Var RegionSize -> Exp2 -> PassM (Exp2, RegionSize)
```

Comment:
The overall structure of this pass looks good. So is the next step here 1) keep a track of which locations belong to which regions, 2) track allocations to locations and update the bounds for a region? 

Also, I think you'll need to know which regions are "input" regions for functions at their call sites. That is, even if a region `r` is passed to a function `f`, if it's an input region for `f`, it'll have 0 allocations there! This comes up while inferring the bounds for `reverse` right?

Reply:
Cool, I'm working on it

Reply:
🎉